File: src/lib/files.rb
# frozen_string_literal: true

require 'pathname'
require 'time'

module Blog
  # Functions for working with files
  module Files
    # Returns the root of the blog project.  Should be something like
    # ~/src/blog
    def self.root
      File.expand_path(File.join(File.dirname(__FILE__), '../..'))
    end

    # Like File.join, except relative to the blog root.  Returns the blog
    # root if called withour args.
    def self.join(*args)
      rel = File.join(root, *args)
      File.expand_path rel
    end

    # Returns a list of entries.
    def self.entries
      @entries ||= Entry.build_paginated_list
    end

    # Returns a list of pages.
    def self.pages
      Dir.glob(Files.join('pages/*.html')).map { |f| Page.new f }
    end

    # Returns a list of ruby test files.
    def self.tests
      Dir.glob(Files.join('src/spec/test_*.rb'))
    end

    # Shorten a filepath relative to the root
    def self.shorten(path)
      path.delete_prefix("#{root}/")
    end

    # Generates a file.  Expects a path and a block that returns a
    # string.  Will report execution time and size.
    def self.generate(path, &block)
      size = 0
      elapsed = Run.time_it do
        content = yield block
        size = content.bytesize
        File.open(path, 'w') do |file|
          file.write(content)
        end
      end
      time = "#{elapsed.round(2)}s"
      Log.info "generated #{shorten(path)} (#{size} b) [#{time}]-> #{shorten(path)}"
    end

    # Joins a file path relative to the site root (www/ by default).
    def self.target(*args)
      join('www/', *args)
    end
  end
end

File: src/lib/files.py
import os
import glob


def root():
    here = os.path.dirname(os.path.realpath(__file__))
    return os.path.abspath(os.path.join(here, '../..'))


def join(*args):
    return os.path.join(root(), *args)


def target(*args):
    return join('www/', *args)


class Page(object):
    def __init__(self, source):
        self.source = source

    def __repr__(self):
        return f'<Page {self.name}>'

    @property
    def name(self):
        filename = os.path.basename(self.source)
        return os.path.splitext(filename)[0] + '.html'

    def generate(self, context):
        pass


class Entry(Page):
    def __repr__(self):
        return f'<Entry {self.name}>'


def pages():
    paths = sorted(glob.glob(join('pages/*.*')), reverse=True)
    return [Page(p) for p in paths]


def entries():
    paths = sorted(glob.glob(join('entries/*.*')), reverse=True)
    return [Entry(p) for p in paths]
